readme.md
# A collection of VCV Rack modules based on audio plugins by Chris Johnson from Airwindows

![Rackwindows Modules](res/images/rackwindows_panels.jpg)

Most of these modules are straight-up ports of [Airwindows](http://www.airwindows.com) plugins with the addition of a panel (designed to be potentially buildable in hardware) and the occasional tweak or enhancement to take advantage of the modular environment.

Airwindows plugins are known for exceptional sound quality, innovative approaches and minimalist interfaces. They often focus on and excel at one specific task, which is why I think they are a natural fit for Rack.

For more in-depth information about the inner workings of a particular Airwindows plugin please check Chris' [website](http://www.airwindows.com). There is a video on every plugin along with a description and I encourage everyone to explore his body of work. Chances are you stumble upon other gems here and there. He's got a [Patreon](https://www.patreon.com/airwindows) in case you appreciate and want to support what he does.

**NOTE:** Most modules act a bit differently depending on the selected sample rate, e.g. varying center frequencies on filters. Keep in mind that a number of Airwindows plugins are actually sample rate dependent, therefore in some cases it might be a feature, not a bug.

## Licence

See [LICENSE.md](LICENSE.md) for all licenses

## Modules

- [Capacitor Mono/Stereo](#capacitor): Filters
- [Chorus](#chorus): Chorus with multi-tap option (Ensemble)
- [Console](#console): Stereo summing mixer
- [Console MM](#consolemm): Stereo summing mixer to work in conjunction with MindMeld's MixMaster
- [Distance](#distance): Designed to mimic through-air high frequency attenuation
- [Dual BSG](#dual-bsg): Dual gain shifter
- [Golem](#golem): Micro-delayable crossfader
- [Holt](#holt): Resonant lowpass filter focussed on low frequencies
- [Hombre](#hombre): Texas tone and texture
- [Interstage](#interstage): Subtle analogifier
- [Monitoring](#monitoring): Mix checker
- [MV](#mv): Dual-mono reverb
- [Rasp](#rasp): De-Edger, slew and acceleration limiter
- [ResEQ](#reseq): Resonance equalizer
- [Tape](#tape): All-purpose tape mojo
- [Tremolo](#tremolo): Organic pulsating thing
- [Vibrato](#vibrato): FM Vibrator o_O

## Capacitor <a id="capacitor"></a>

Filters

![Rackwindows Capacitor](res/images/capacitor_panels.jpg)

High/Lowpass filters that come in both mono and stereo, with the stereo version featuring an additional dry/wet control.

[More information](http://www.airwindows.com/capacitor)

## Chorus <a id="chorus"></a>

Chorus with multi-tap option (Ensemble) 

![Rackwindows Chorus](res/images/chorus_panels.jpg)

This surprisingly versatile module is a combination of Airwindows Chorus and ChorusEnsemble. The switch lets you toggle between the two flavours.

[More information](http://www.airwindows.com/chorus-vst)

## Console <a id="console"></a>

Stereo summing mixer

![Rackwindows Console](res/images/console_panels.jpg)

Airwindows Console systems sum signals in a way where the resulting soundstage appears less flat, more spacious and more in line with what we are used to from analog mixers. There are a number of variations/flavours of the console concept and this module currently implements the most recent one, Console6, and PurestConsole, which can be selected in the context menu. 

Please check the link below for further information on how exactly this effect is achieved.

**NOTE**: Due to how the encoding/decoding works, spaciousness and definition will increase with each additional channel. There will be **no effect on a single channel** at all.

[More information](http://www.airwindows.com/console2)

## Console MM <a id="consolemm"></a>

This is a special version of Console designed to work alongside and interconnected with MindMeld's MixMaster module. It takes MM's polyphonic direct outs and sums the individual channels according to the selected console type. The stereo outputs can then be routed back into the chain inputs of MM to complete the roundtrip (make sure to check "Solo Chain Input" in MM, rightclick on Master). This setup conveniently provides "analogish" summing while making use of all of MM's fantastic mixing capabilities and goodies.

For added flexibility and in order to honour Chris' spirit of allowing for settings beyond obvious sweetspots, the input gain can be adjusted, or - as a friend put it - taken from "can`t hear shit happening" to overdriven. 

The 3 direct outputs pass on the respective input signal either unprocessed (polyphonic) or summed (mono).

## Distance <a id="distance"></a>

Designed to mimic through-air high frequency attenuation

![Rackwindows Distance](res/images/distance_panels.jpg)

Pushes things back. Also quite nice on reverb returns.

[More information](http://www.airwindows.com/distance-vst)

## Dual BSG <a id="dual-bsg"></a>

Dual gain shifter

![Rackwindows Dual BSG](res/images/dual_bsg_panels.jpg)

Scales a signal up or down by increments of exactly 6 dB. If no input is connected, the respective output will provide constant voltage selectable in 1V steps from -8V to +8V. The lower section can be linked to the upper one to automatically compensate for values set by the upper 'Shift' knob. If linked the lower 'Shift' knob can be used to offset the signal in 6db steps (input connected) or 1V steps (input not connected). 

[More information](http://www.airwindows.com/bitshiftgain)

## Golem <a id="golem"></a>

Micro-delayable crossfader

![Rackwindows Golem](res/images/golem_panels.jpg)

[More information](https://www.airwindows.com/golem-vst/)

## Holt <a id="holt"></a>

Resonant lowpass filter focussed on low frequencies

![Rackwindows Holt](res/images/holt_panels.jpg)

Interestingly, Holt's algorithm is based on an Excel method for predicting sales figures based on trends. The result is a lowpass filter that's polite in the highs and increasingly mean towards the low-end. It also allows for seamless morphing between no poles (dry) and 4-poles (24db per octave). An additional output saturation stage helps keeping the possibly massive resonances in check (it can still get nasty though, you have been warned).

[More information](http://www.airwindows.com/holt)

## Hombre <a id="hombre"></a>

Texas tone and texture

![Rackwindows Hombre](res/images/hombre_panels.jpg)

[More information](http://www.airwindows.com/hombre-vst)

## Interstage <a id="interstage"></a>

Subtle analogifier

![Rackwindows Interstage](res/images/interstage_panels.jpg)

Well, I won't attempt to describe or even explain what this one does. Please head straight over to the Airwindows site for details. But I will say this: If I had to choose just one Airwindows tool, this would be it (or maybe Console?). It's simple, it's subtle and to my ears pure class. Put it anywhere in your patch, you can do no wrong. It will most likely be for the better :-)

[More information](https://www.airwindows.com/interstage/)

## Monitoring <a id="monitoring"></a>

Mix checker

![Rackwindows Monitoring](res/images/monitoring_panels.jpg)

Monitoring is meant to be the last module before the output. It offers a number of tools to help investigate your patch and check for problems. It is calibrated in such a way that modes like *Subs* or *Peaks* will tell you unmistakably when you've got too much (or not enough) energy in a particular area.
Additionally, it features crossfeed for headphones and optional 24/16 bit dither.

### Processing modes:

- **Subs**: This is [SubsOnly](https://www.airwindows.com/slewonly-subsonly/). Useful for checking sub-bass elements.
- **Slew**: This is [SlewOnly](https://www.airwindows.com/slewonly-subsonly/). Similar to *Subs*, but for high frequency content.
- **Peaks**: This is [PeaksOnly](https://www.airwindows.com/peaksonly/). Exposes bursts of excessive energy by transforming them into recognizable signals.
- **Mono**: Mid or sum of the stereo channels.
- **Side**: Just the side information or difference of the stereo channels.
- **Vinyl**: Rolls off the extreme ends of the frequency spectrum.
- **Aurat**: Same as Vinyl, but slightly narrower band.
- **Phone**: Same as above, but significantly less bass and mono.

### Can modes:

These are somewhat sophisticated crossfeed modes for headphones. They go from relatively subtle (A) to quite obvious (D). 

### Dither:

You can choose between 24 and 16 bit dithering. The algorithm is [Dark](https://www.airwindows.com/dark/).

[More information](https://www.airwindows.com/monitoring/)

## MV <a id="mv"></a>

Dual-mono reverb

![Rackwindows MV](res/images/mv_panels.jpg)

A reverb based on Bitshiftgain and old Alesis Midiverbs. Capable of turning everything into a pad or sustaining a 'bloom' forever. Watch your volume when using small amounts of depth with full on regeneration.

[More information](http://www.airwindows.com/mv)

## Rasp <a id="rasp"></a>

De-edger, high frequency tamer, acceleration limiter

![Rackwindows Rasp](res/images/rasp_panels.jpg)

Rasp combines Slew/Slew2/Slew3 and Acceleration in a single module for versatile high frequency conditioning. The Slew algorithms are significantly different in character and can be selected in the menu.

**Routing:** If only the *Limit* output is connected, Slew is routed into Acceleration. If only the *Clamp* output is connected, Acceleration runs into Slew. If both outputs are connected, *Clamp* outputs Slew and *Limit* outputs Acceleration.

[More information on Acceleration](https://www.airwindows.com/acceleration/)
[Slew](https://www.airwindows.com/slew-2/), [Slew2](https://www.airwindows.com/slew2-2/), [Slew3](https://www.airwindows.com/slew3/)

## ResEQ <a id="reseq"></a>

Resonance equalizer

![Rackwindows ResEQ](res/images/reseq_panels.jpg)

ResEQ passes audio through up to four adjustable frequency bands - and only those bands. All bands are similar and with identical range, but interact with each other in a particular way. They can be stacked for massive boosts or thin the signal out if set slightly apart. What sounds artificial in isolation, can be useful to highlight certain characteristics of a sound and then blend to taste with the dry/wet for more natural results.

[More information](https://www.airwindows.com/reseq-vst/)

## Tape <a id="tape"></a>

All-purpose tape mojo

![Rackwindows Tape](res/images/tape_panels.jpg)

Tape is Airwindows' attempt to capture the very essence of what running audio through tape machines does to the signal. Driving it not only alters the volume, but also applies certain characteristics which are often associated with tape. The bump control allows to adjust the low end.

[More information](http://www.airwindows.com/tape)

## Tremolo <a id="tremolo"></a>

Organic pulsating thing

![Rackwindows Tremolo](res/images/tremolo_panels.jpg)

Instead of volume animation Tremolo uses fluctuating saturation and antisaturation curves to create a more organic, somewhat tubey tremolo effect.

[More information](http://www.airwindows.com/tremolo-vst)

## Vibrato <a id="vibrato"></a>

FM Vibrator o_O

![Rackwindows Vibrato](res/images/vibrato_panels.jpg)

Make sure to also play with the 'Inverse/Wet' knob for chorusing and flange effects. Two trigger outputs have been added for fun.

[More information](http://www.airwindows.com/vibrato-vst)

## A word on processing quality

Most modules feature an **Eco** mode in order to reduce CPU usage on weaker systems. The actual algorithms remain untouched, but any noise shaping/dithering is skipped. This can result in speed improvements of roughly 10% to 50% depending on the module.

## Building from Source

To compile the modules from source, see the official [VCV Rack documentation](https://vcvrack.com/manual/Building.html).

## Colophon

The typeface used on the panels is [Barlow](https://github.com/jpt/barlow) by Jeremy Tribby.

changelog.md

### 1.1.2 (13-09-2020)
- New module: Console MM
- New module: Monitoring
- New module: Golem
- New module: Rasp
- Tape: Fixed unbalanced output when loaded in a patch

### 1.1.1 (29-07-2020)
- New module: Holt
- ResEQ: Fixed parameter scaling and glitches
- Tape: Improved performance
- DualBSG: Minimized clicks on parameter changes (monophonic only)
- Hombre: Removed wrong tag
- Some modules: Fixed a bug, where the last voltage would be output after disconnection of input
- All: Revised initialization and reset behaviour

### 1.1.0 (05-07-2020)
- New module: ResEQ
- New module: Interstage
- Added/Updated low CPU mode for Chorus, Console, Distance, Hombre, MV, Tape, Tremolo and Vibrato
- All: minor performance optimizations

### 1.0.4 (22-05-2020)
- Now polyphonic: Dual BSG, Capacitor, Capacitor ST, Chorus, Console, Distance, Hombre, Tape, Tremolo and Vibrato
- Capacitor, Capacitor ST: added low cpu mode

### 1.0.3 (24-04-2020)

- New module: Tape
- New module: Console

### 1.0.2 (13-04-2020)

- Dual BSG: fixed switch graphic not rendering properly

### 1.0.1 (07-04-2020)

- New module: Capacitor Stereo
- New module: Distance
- New module: Tremolo

- All: process audio only when an output is connected
- All: minor fixes and code cleanup

### 1.0.0 (27-03-2020)

- Initial release 

src/reseq.cpp
/***********************************************************************************************
ResEQ
------
VCV Rack module based on ResEQ by Chris Johnson from Airwindows <www.airwindows.com>

Ported and designed by Jens Robert Janke 

Changes/Additions:
- 4 frequency beams instead of 8
- mono
- CV inputs for frequency beams and drywet
- polyphonic

See ./LICENSE.md for all licenses
************************************************************************************************/

#include "plugin.hpp"

// quality options
#define ECO 0
#define HIGH 1

struct Reseq : Module {
    enum ParamIds {
        ENUMS(RESO_PARAMS, 4),
        DRYWET_PARAM,
        NUM_PARAMS
    };
    enum InputIds {
        ENUMS(RESO_CV_INPUTS, 4),
        DRYWET_CV_INPUT,
        IN_INPUT,
        NUM_INPUTS
    };
    enum OutputIds {
        OUT_OUTPUT,
        NUM_OUTPUTS
    };
    enum LightIds {
        NUM_LIGHTS
    };

    // module variables
    const double gainFactor = 32.0;
    int quality;

    // control parameters
    float r1Param;
    float r2Param;
    float r3Param;
    float r4Param;
    float drywetParam;
    bool isActiveR1;
    bool isActiveR2;
    bool isActiveR3;
    bool isActiveR4;

    // state variables (as arrays in order to handle up to 16 polyphonic channels)
    double b[16][61];
    double f[16][61];
    int framenumber[16];
    uint32_t fpd[16];

    // other variables, which do not need to be updated every cycle
    double overallscale;
    double v1;
    double v2;
    double v3;
    double v4;
    double f1;
    double f2;
    double f3;
    double f4;
    double wet;
    double falloff;

    Reseq()
    {
        config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
        for (int i = 0; i < 4; i++) {
            configParam(RESO_PARAMS + i, 0.f, 1.f, 0.f, string::f("Reso %d", i + 1), "%", 0, 100);
        }
        configParam(DRYWET_PARAM, 0.f, 1.f, 1.f, "Dry/Wet");

        quality = loadQuality();
        onReset();
    }

    void onReset() override
    {
        onSampleRateChange();
        updateParams();

        isActiveR1 = isActiveR2 = isActiveR3 = isActiveR4 = false;

        for (int i = 0; i < 16; i++) {
            for (int x = 0; x < 61; x++) {
                b[i][x] = 0.0;
                f[i][x] = 0.0;
            }
            framenumber[i] = 1;
            fpd[i] = 17;
        }
    }

    void onSampleRateChange() override
    {
        float sampleRate = APP->engine->getSampleRate();

        overallscale = 1.0;
        overallscale /= 44100.0;
        overallscale *= sampleRate;
    }

    json_t* dataToJson() override
    {
        json_t* rootJ = json_object();

        // quality
        json_object_set_new(rootJ, "quality", json_integer(quality));

        return rootJ;
    }

    void dataFromJson(json_t* rootJ) override
    {
        // quality
        json_t* qualityJ = json_object_get(rootJ, "quality");
        if (qualityJ)
            quality = json_integer_value(qualityJ);
    }

    void updateParams()
    {
        r1Param = params[RESO_PARAMS + 0].getValue();
        r1Param += inputs[RESO_CV_INPUTS + 0].getVoltage() / 5;
        r1Param = clamp(r1Param, 0.f, 1.f);

        r2Param = params[RESO_PARAMS + 1].getValue();
        r2Param += inputs[RESO_CV_INPUTS + 1].getVoltage() / 5;
        r2Param = clamp(r2Param, 0.f, 1.f);

        r3Param = params[RESO_PARAMS + 2].getValue();
        r3Param += inputs[RESO_CV_INPUTS + 2].getVoltage() / 5;
        r3Param = clamp(r3Param, 0.f, 1.f);

        r4Param = params[RESO_PARAMS + 3].getValue();
        r4Param += inputs[RESO_CV_INPUTS + 3].getVoltage() / 5;
        r4Param = clamp(r4Param, 0.f, 1.f);

        drywetParam = params[DRYWET_PARAM].getValue();
        drywetParam += inputs[DRYWET_CV_INPUT].getVoltage() / 5;
        drywetParam = clamp(drywetParam, 0.f, 1.f);

        wet = drywetParam;

        if (r1Param) {
            v1 = r1Param;
            f1 = pow(v1, 2);
            v1 += 0.2;
            v1 /= overallscale;
            isActiveR1 = true;
        } else {
            isActiveR1 = false;
        }

        if (r2Param) {
            v2 = r2Param;
            f2 = pow(v2, 2);
            v2 += 0.2;
            v2 /= overallscale;
            isActiveR2 = true;
        } else {
            isActiveR2 = false;
        }

        if (r3Param) {
            v3 = r3Param;
            f3 = pow(v3, 2);
            v3 += 0.2;
            v3 /= overallscale;
            isActiveR3 = true;
        } else {
            isActiveR3 = false;
        }

        if (r4Param) {
            v4 = r4Param;
            f4 = pow(v4, 2);
            v4 += 0.2;
            v4 /= overallscale;
            isActiveR4 = true;
        } else {
            isActiveR4 = false;
        }
    }

    void processChannel(Input& input, Output& output)
    {
        // number of polyphonic channels
        int numChannels = std::max(1, inputs[IN_INPUT].getChannels());

        // for each poly channel
        for (int i = 0; i < numChannels; ++i) {

            // input
            long double inputSample = input.getPolyVoltage(i);

            // pad gain
            inputSample /= gainFactor;

            // each process frame we'll update some of the kernel frames. That way we don't have to crunch the whole thing at once,
            // and we can load a LOT more resonant peaks into the kernel.
            framenumber[i] += 1;
            if (framenumber[i] > 59)
                framenumber[i] = 1;
            falloff = sin(framenumber[i] / 19.098992);
            f[i][framenumber[i]] = 0.0;

            if (isActiveR1) {
                if ((framenumber[i] * f1) < 1.57079633)
                    f[i][framenumber[i]] += (sin((framenumber[i] * f1) * 2.0) * falloff * v1);
                else
                    f[i][framenumber[i]] += (cos(framenumber[i] * f1) * falloff * v1);
            }
            if (isActiveR2) {
                if ((framenumber[i] * f2) < 1.57079633)
                    f[i][framenumber[i]] += (sin((framenumber[i] * f2) * 2.0) * falloff * v2);
                else
                    f[i][framenumber[i]] += (cos(framenumber[i] * f2) * falloff * v2);
            }
            if (isActiveR3) {
                if ((framenumber[i] * f3) < 1.57079633)
                    f[i][framenumber[i]] += (sin((framenumber[i] * f3) * 2.0) * falloff * v3);
                else
                    f[i][framenumber[i]] += (cos(framenumber[i] * f3) * falloff * v3);
            }
            if (isActiveR4) {
                if ((framenumber[i] * f4) < 1.57079633)
                    f[i][framenumber[i]] += (sin((framenumber[i] * f4) * 2.0) * falloff * v4);
                else
                    f[i][framenumber[i]] += (cos(framenumber[i] * f4) * falloff * v4);
            }
            //done updating the kernel for this go-round

            if (quality == HIGH) {
                if (fabs(inputSample) < 1.18e-43)
                    inputSample = fpd[i] * 1.18e-43;
            }

            long double drySample = inputSample;

            // EQ kernel
            b[i][59] = b[i][58];
            b[i][58] = b[i][57];
            b[i][57] = b[i][56];
            b[i][56] = b[i][55];
            b[i][55] = b[i][54];
            b[i][54] = b[i][53];
            b[i][53] = b[i][52];
            b[i][52] = b[i][51];
            b[i][51] = b[i][50];
            b[i][50] = b[i][49];
            b[i][49] = b[i][48];
            b[i][48] = b[i][47];
            b[i][47] = b[i][46];
            b[i][46] = b[i][45];
            b[i][45] = b[i][44];
            b[i][44] = b[i][43];
            b[i][43] = b[i][42];
            b[i][42] = b[i][41];
            b[i][41] = b[i][40];
            b[i][40] = b[i][39];
            b[i][39] = b[i][38];
            b[i][38] = b[i][37];
            b[i][37] = b[i][36];
            b[i][36] = b[i][35];
            b[i][35] = b[i][34];
            b[i][34] = b[i][33];
            b[i][33] = b[i][32];
            b[i][32] = b[i][31];
            b[i][31] = b[i][30];
            b[i][30] = b[i][29];
            b[i][29] = b[i][28];
            b[i][28] = b[i][27];
            b[i][27] = b[i][26];
            b[i][26] = b[i][25];
            b[i][25] = b[i][24];
            b[i][24] = b[i][23];
            b[i][23] = b[i][22];
            b[i][22] = b[i][21];
            b[i][21] = b[i][20];
            b[i][20] = b[i][19];
            b[i][19] = b[i][18];
            b[i][18] = b[i][17];
            b[i][17] = b[i][16];
            b[i][16] = b[i][15];
            b[i][15] = b[i][14];
            b[i][14] = b[i][13];
            b[i][13] = b[i][12];
            b[i][12] = b[i][11];
            b[i][11] = b[i][10];
            b[i][10] = b[i][9];
            b[i][9] = b[i][8];
            b[i][8] = b[i][7];
            b[i][7] = b[i][6];
            b[i][6] = b[i][5];
            b[i][5] = b[i][4];
            b[i][4] = b[i][3];
            b[i][3] = b[i][2];
            b[i][2] = b[i][1];
            b[i][1] = b[i][0];
            b[i][0] = inputSample;

            inputSample = (b[i][1] * f[i][1]);
            inputSample += (b[i][2] * f[i][2]);
            inputSample += (b[i][3] * f[i][3]);
            inputSample += (b[i][4] * f[i][4]);
            inputSample += (b[i][5] * f[i][5]);
            inputSample += (b[i][6] * f[i][6]);
            inputSample += (b[i][7] * f[i][7]);
            inputSample += (b[i][8] * f[i][8]);
            inputSample += (b[i][9] * f[i][9]);
            inputSample += (b[i][10] * f[i][10]);
            inputSample += (b[i][11] * f[i][11]);
            inputSample += (b[i][12] * f[i][12]);
            inputSample += (b[i][13] * f[i][13]);
            inputSample += (b[i][14] * f[i][14]);
            inputSample += (b[i][15] * f[i][15]);
            inputSample += (b[i][16] * f[i][16]);
            inputSample += (b[i][17] * f[i][17]);
            inputSample += (b[i][18] * f[i][18]);
            inputSample += (b[i][19] * f[i][19]);
            inputSample += (b[i][20] * f[i][20]);
            inputSample += (b[i][21] * f[i][21]);
            inputSample += (b[i][22] * f[i][22]);
            inputSample += (b[i][23] * f[i][23]);
            inputSample += (b[i][24] * f[i][24]);
            inputSample += (b[i][25] * f[i][25]);
            inputSample += (b[i][26] * f[i][26]);
            inputSample += (b[i][27] * f[i][27]);
            inputSample += (b[i][28] * f[i][28]);
            inputSample += (b[i][29] * f[i][29]);
            inputSample += (b[i][30] * f[i][30]);
            inputSample += (b[i][31] * f[i][31]);
            inputSample += (b[i][32] * f[i][32]);
            inputSample += (b[i][33] * f[i][33]);
            inputSample += (b[i][34] * f[i][34]);
            inputSample += (b[i][35] * f[i][35]);
            inputSample += (b[i][36] * f[i][36]);
            inputSample += (b[i][37] * f[i][37]);
            inputSample += (b[i][38] * f[i][38]);
            inputSample += (b[i][39] * f[i][39]);
            inputSample += (b[i][40] * f[i][40]);
            inputSample += (b[i][41] * f[i][41]);
            inputSample += (b[i][42] * f[i][42]);
            inputSample += (b[i][43] * f[i][43]);
            inputSample += (b[i][44] * f[i][44]);
            inputSample += (b[i][45] * f[i][45]);
            inputSample += (b[i][46] * f[i][46]);
            inputSample += (b[i][47] * f[i][47]);
            inputSample += (b[i][48] * f[i][48]);
            inputSample += (b[i][49] * f[i][49]);
            inputSample += (b[i][50] * f[i][50]);
            inputSample += (b[i][51] * f[i][51]);
            inputSample += (b[i][52] * f[i][52]);
            inputSample += (b[i][53] * f[i][53]);
            inputSample += (b[i][54] * f[i][54]);
            inputSample += (b[i][55] * f[i][55]);
            inputSample += (b[i][56] * f[i][56]);
            inputSample += (b[i][57] * f[i][57]);
            inputSample += (b[i][58] * f[i][58]);
            inputSample += (b[i][59] * f[i][59]);
            inputSample /= 12.0;
            //inlined- this is our little EQ kernel. Longer will give better tightness on bass frequencies.
            //Note that normal programmers will make this a loop, which isn't much slower if at all, on modern CPUs.
            //It was unrolled more or less to show how much is done when you define a loop like that: it's easy to specify stuff where a lot of grinding is done.
            //end EQ kernel

            if (wet != 1.0) {
                inputSample = (inputSample * wet) + (drySample * (1.0 - wet));
            }

            if (quality == HIGH) {
                //begin 64 bit stereo floating point dither
                int expon;
                frexp((double)inputSample, &expon);
                fpd[i] ^= fpd[i] << 13;
                fpd[i] ^= fpd[i] >> 17;
                fpd[i] ^= fpd[i] << 5;
                inputSample += ((double(fpd[i]) - uint32_t(0x7fffffff)) * 1.1e-44l * pow(2, expon + 62));
                //end 64 bit stereo floating point dither
            }

            // bring gain back up
            inputSample *= gainFactor;

            // output
            outputs[OUT_OUTPUT].setChannels(numChannels);
            outputs[OUT_OUTPUT].setVoltage(inputSample, i);

        } // end poly channel loop
    }

    void process(const ProcessArgs& args) override
    {
        if (outputs[OUT_OUTPUT].isConnected()) {

            updateParams();
            processChannel(inputs[IN_INPUT], outputs[OUT_OUTPUT]);
        }
    }
};

struct ReseqWidget : ModuleWidget {

    // quality item
    struct QualityItem : MenuItem {
        Reseq* module;
        int quality;

        void onAction(const event::Action& e) override
        {
            module->quality = quality;
        }

        void step() override
        {
            rightText = (module->quality == quality) ? "✔" : "";
        }
    };

    void appendContextMenu(Menu* menu) override
    {
        Reseq* module = dynamic_cast<Reseq*>(this->module);
        assert(module);

        menu->addChild(new MenuSeparator()); // separator

        MenuLabel* qualityLabel = new MenuLabel(); // menu label
        qualityLabel->text = "Quality";
        menu->addChild(qualityLabel);

        QualityItem* low = new QualityItem(); // low quality
        low->text = "Eco";
        low->module = module;
        low->quality = 0;
        menu->addChild(low);

        QualityItem* high = new QualityItem(); // high quality
        high->text = "High";
        high->module = module;
        high->quality = 1;
        menu->addChild(high);
    }

    ReseqWidget(Reseq* module)
    {
        setModule(module);
        setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/reseq_dark.svg")));

        addChild(createWidget<ScrewBlack>(Vec(RACK_GRID_WIDTH, 0)));
        addChild(createWidget<ScrewBlack>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, 0)));
        addChild(createWidget<ScrewBlack>(Vec(RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));
        addChild(createWidget<ScrewBlack>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));

        addParam(createParamCentered<RwKnobSmallDark>(Vec(22.5, 55.0), module, Reseq::RESO_PARAMS + 0));
        addParam(createParamCentered<RwKnobSmallDark>(Vec(67.5, 55.0), module, Reseq::RESO_PARAMS + 1));
        addParam(createParamCentered<RwKnobSmallDark>(Vec(22.5, 105.0), module, Reseq::RESO_PARAMS + 2));
        addParam(createParamCentered<RwKnobSmallDark>(Vec(67.5, 105.0), module, Reseq::RESO_PARAMS + 3));
        addParam(createParamCentered<RwKnobMediumDark>(Vec(45.0, 155.0), module, Reseq::DRYWET_PARAM));

        addInput(createInputCentered<RwPJ301MPortSilver>(Vec(26.25, 245.0), module, Reseq::RESO_CV_INPUTS + 0));
        addInput(createInputCentered<RwPJ301MPortSilver>(Vec(63.75, 245.0), module, Reseq::RESO_CV_INPUTS + 1));
        addInput(createInputCentered<RwPJ301MPortSilver>(Vec(26.25, 285.0), module, Reseq::RESO_CV_INPUTS + 2));
        addInput(createInputCentered<RwPJ301MPortSilver>(Vec(63.75, 285.0), module, Reseq::RESO_CV_INPUTS + 3));
        addInput(createInputCentered<RwPJ301MPortSilver>(Vec(45.0, 205.0), module, Reseq::DRYWET_CV_INPUT));
        addInput(createInputCentered<RwPJ301MPortSilver>(Vec(26.25, 325.0), module, Reseq::IN_INPUT));

        addOutput(createOutputCentered<RwPJ301MPort>(Vec(63.75, 325.0), module, Reseq::OUT_OUTPUT));
    }
};

Model* modelReseq = createModel<Reseq, ReseqWidget>("reseq");